db/python/tables/participant.py

```diff
@@ -249,11 +323,13 @@ async def get_participants_by_families(
         self, family_ids: list[int]
     ) -> tuple[set[ProjectId], dict[int, list[ParticipantInternal]]]:
         """Get list of participants keyed by families, duplicates results"""
-        _query = """
-            SELECT project, fp.family_id, p.id, p.external_id, p.reported_sex, p.reported_gender, p.karyotype, p.meta
+        _query = f"""
+            SELECT fp.family_id, {self.keys_str}
             FROM participant p
             INNER JOIN family_participant fp ON fp.participant_id = p.id
+            INNER JOIN participant_external_id peid ON p.project = peid.project AND p.id = peid.participant_id
```

Comment:
I don't think you need the project check here, I don't know if it has any project implications though. I guess it opens up can a participant have external IDs outside their project? Sounds a little dangerous

Reply:
Yes, `p.id` is universally unique, so just checking that would probably be fine.

The backstory is I put a project column in the participant_external_id table because #139 listed one, and eventually because that makes it easy to wipe a project out of the table in `delete_project_data()`.

I thought I had the primary key as project+participant_id and thus there was some cargo-cult programming going on along these lines: therefore there is an index on project+participant_id, therefore it will be most efficient to do the join on the components of the index.

However actually the primary key is participant_id+name (each participant can only have one external_id of each flavour), and there is a unique constraint on project+name+external_id (each external_id must be unique within its flavour within each project).  (Usually I say “external organisation” rather than “flavour”!)

So this is all a bit artificial.

Maybe the family_participant table points to the right answer… Maybe the participant_external_id table shouldn't actually have a `project` column at all?! It doesn't really need one, as participant_id uniquely points to a particular participant regardless of project. Particular implications would be:

1. Slightly harder to wipe out on a per-project basis in `delete_project_data()`, but again family_participant shows how to do that using a sub-select.
2. That unique constraint would just be on name+external_id, so external_ids for a particular external org would need to be unique universally rather than within a particular project. That's a bit bad in that it reduces project independence, but in practice probably doesn't matter at all…
3. I wouldn't need to do the stupid query in the `to_update` part of `ParticipantTable.update_participant()` to figure out what the project is. :+1:
4. I guess not duplicating `project` down here would mean the table was better normalised. :+1:

Reply:
I prepared the commit of not having a `project` column in these two *_external_id tables, and I think I like it: see this [mult-extids-no-project branch](https://github.com/populationgenomics/metamist/compare/multiple-sample-extids...mult-extids-no-project).

Interested in your thoughts @illusional, on this and on whether having a “external_ids must be unique within each external org” constraint is worthwhile.

Reply:
I've removed the project checks on all those JOINs, and removed the now superseded `project,external_id` constraints on the participant and sample tables.

Thus it turns out that we were indeed previously enforcing “external ids must be unique within a project”, so (apropos (2) above) we should indeed continue to enforce “external ids for the same org name must be unique **within a project**” — hence it is worth keeping the `project` column in these tables.

(Just for posterity: @illusional Was that your reasoning for disliking removing these `project` columns, or were there other aspects as well?)

Reply:
Now addressed: the tables retain their `project` columns, and `name` no longer contributes to the uniqueness constraint.